File: 3432-count-partitions-with-even-sum-difference.py
class Solution:
    def countPartitions(self, nums: list[int]) -> int:
        '''
        We first get the total sum of the nums
        We go through each index,
            - at each `i`,
                - We can have a cur_total var that calculates the total up till this `i` index
                - then we calculate the sum of right subarray by subtracting `cur_total` from `total`
                - next, we calculate the difference between these 2 and if it's even, we add 1 to the `res` count

        '''
        total = sum(nums)
        cur_total = 0
        res = 0
        for i in range(0, len(nums) - 1):
            cur_total += nums[i]
            right_total = total - cur_total
            if (cur_total - right_total) % 2 == 0:
                res += 1

        return res


sol = Solution()
print(f'output: {sol.countPartitions([10, 10, 3, 7, 6])}, expected: 4')
print(f'output: {sol.countPartitions([1, 2, 2])}, expected: 0')
print(f'output: {sol.countPartitions([2, 4, 6, 8])}, expected: 3')

File: 3432-count-partitions-with-even-sum-difference.java
class CountPartitionsWithEvenSumDifference {
    public int countPartitions(int[] nums) {
        int total = 0;
        for (int num : nums) {
            total += num;
        }
        int res = 0;
        int curTotal = 0;
        for (int i = 0; i < nums.length - 1; i++) {
            curTotal += nums[i];
            if ((2 * curTotal - total) % 2 == 0) {
                res++;
            }
        }

        return res;
    }
  
    public static void main(String[] args) {
        CountPartitionsWithEvenSumDifference sol = new CountPartitionsWithEvenSumDifference();
        System.out.println("Output: " + sol.countPartitions(new int[]{10,10,3,7,6}) + ", expected: 4");
        System.out.println("Output: " + sol.countPartitions(new int[]{1,2,2}) + ", expected: 0");
        System.out.println("Output: " + sol.countPartitions(new int[]{2,4,6,8}) + ", expected: 3");
    }
}
